### PR TITLE
No array access in strings

### PR DESCRIPTION
**Pull request by bizley**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | -

### CODE REVIEW

**Comment by bizley** [comment]
To fix altogether cases like #18395, #18426, #9754

File: build/controllers/PhpDocController.php [hunk 1]
```diff
@@ -329,16 +329,16 @@ protected function fixDocBlockIndentation(&$lines)
                 $tag = false;
             } elseif ($docBlock) {
                 $line = ltrim($line);
-                if (isset($line[0]) && $line[0] === '*') {
+                if (!empty($line) && strpos($line, '*') === 0) {
                     $line = substr($line, 1);
                 }
-                if (isset($line[0]) && $line[0] === ' ') {
+                if (!empty($line) && strpos($line, ' ') === 0) {
```
Comment (samdark) [inline build/controllers/PhpDocController.php:335]:
```suggestion
                if (strpos($line, ' ') === 0) {
```

File: build/controllers/PhpDocController.php [hunk 2]
```diff
@@ -329,16 +329,16 @@ protected function fixDocBlockIndentation(&$lines)
                 $tag = false;
             } elseif ($docBlock) {
                 $line = ltrim($line);
-                if (isset($line[0]) && $line[0] === '*') {
+                if (!empty($line) && strpos($line, '*') === 0) {
```
Comment (samdark) [inline build/controllers/PhpDocController.php:332]:
```suggestion
                if (strpos($line, '*') === 0) {
```

File: build/controllers/PhpDocController.php [hunk 3]
```diff
@@ -453,15 +453,15 @@ protected function fixLineSpacing($lines)
...
-            } elseif (substr($line, 0, 6) === 'const ') {
+            } elseif (strpos($line, 'const ') === 0) {
                 $endofConst = $i;
                 $property = false;
-            } elseif (substr($line, 0, 4) === 'use ') {
+            } elseif (strpos($line, 'use ') === 0) {
                 $endofUse = $i;
                 $property = false;
-            } elseif (!empty($line) && $line[0] === '*') {
+            } elseif (!empty($line) && strpos($line, '*') === 0) {
```
Comment (samdark) [inline build/controllers/PhpDocController.php:462]:
```suggestion
            } elseif (strpos($line, '*') === 0) {
```

File: build/controllers/PhpDocController.php [hunk 4]
```diff
@@ -453,15 +453,15 @@ protected function fixLineSpacing($lines)
...
                 $property = false;
-            } elseif (!empty($line) && $line[0] !== '*' && strpos($line, 'function ') !== false || $line === '}') {
+            } elseif (!empty($line) && strpos($line, '*') !== 0 && strpos($line, 'function ') !== false || $line === '}') {
```
Comment (samdark) [inline build/controllers/PhpDocController.php:464]:
```suggestion
            } elseif (strpos($line, '*') !== 0 && strpos($line, 'function ') !== false || $line === '}') {
```

File: build/controllers/TranslationController.php [hunk 5]
```diff
@@ -126,11 +126,11 @@ public function highlightDiff($diff)
     {
         $lines = explode("\n", $diff);
         foreach ($lines as $key => $val) {
-            if (mb_substr($val, 0, 1, 'utf-8') === '@') {
+            if (strpos($val, '@') === 0) {
```
Comment (samdark) [inline build/controllers/TranslationController.php:129]: Should it be `mb_strpos`?

File: framework/console/controllers/MessageController.php [hunk 6]
```diff
@@ -386,7 +386,7 @@ protected function saveMessagesToDb($messages, $db, $sourceMessageTable, $messag
 
         if (!$removeUnused) {
             foreach ($obsolete as $pk => $msg) {
-                if (mb_substr($msg, 0, 2) === '@@' && mb_substr($msg, -2) === '@@') {
+                if (strpos($msg, '@@') === 0 && substr($msg, -2) === '@@') {
```
Comment (samdark) [inline framework/console/controllers/MessageController.php:389]: I think these shoud be kept multibyte-aware.

File: framework/db/pgsql/Schema.php [hunk 7]
```diff
@@ -443,7 +443,7 @@ public function findUniqueIndexes($table)
                 $row = array_change_key_case($row, CASE_LOWER);
             }
             $column = $row['columnname'];
-            if (!empty($column) && $column[0] === '"') {
+            if (!empty($column) && strpos($column, '"') === 0) {
```
Comment (samdark) [inline framework/db/pgsql/Schema.php:446]:
```suggestion
            if (strpos($column, '"') === 0) {
```

File: framework/helpers/BaseFileHelper.php [hunk 8]
```diff
@@ -700,7 +700,7 @@ private static function matchBasename($baseName, $pattern, $firstWildcard, $flag
     private static function matchPathname($path, $basePath, $pattern, $firstWildcard, $flags)
     {
         // match with FNM_PATHNAME; the pattern has base implicitly in front of it.
-        if (isset($pattern[0]) && $pattern[0] === '/') {
+        if (!empty($pattern) && strpos($pattern, '/') === 0) {
```
Comment (samdark) [inline framework/helpers/BaseFileHelper.php:703]:
```suggestion
        if (strpos($pattern, '/') === 0) {
```

File: tests/framework/helpers/FileHelperTest.php [hunk 9]
```diff
@@ -396,7 +396,7 @@ public function testRemoveDirectory()
 
     public function testRemoveDirectorySymlinks1()
     {
-        if (strtolower(substr(PHP_OS, 0, 3)) == 'win') {
+        if (stripos(PHP_OS, 'win') === 0) {
```
Comment (samdark) [inline tests/framework/helpers/FileHelperTest.php:399]:
```suggestion
        if (DIRECTORY_SEPARATOR === '\\') {
```

File: tests/framework/helpers/FileHelperTest.php [hunk 10]
```diff
@@ -439,7 +439,7 @@ public function testRemoveDirectorySymlinks1()
 
     public function testRemoveDirectorySymlinks2()
     {
-        if (strtolower(substr(PHP_OS, 0, 3)) == 'win') {
+        if (stripos(PHP_OS, 'win') === 0) {
```
Comment (samdark) [inline tests/framework/helpers/FileHelperTest.php:442]:
```suggestion
        if (DIRECTORY_SEPARATOR === '\\') {
```

**Review by samdark — CHANGES_REQUESTED** [review]

File: build/controllers/TranslationController.php (hunk 5 above)
Comment (bizley) [reply build/controllers/TranslationController.php:129]: I don't think so. `strpos` is binary-safe and we are looking for ASCII character.

File: framework/console/controllers/MessageController.php (hunk 6 above)
Comment (bizley) [reply framework/console/controllers/MessageController.php:389]: See my comment from above. And I have tested `substr($msg, -2) === '@@'` with some arabic text.

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
👍

**Pull request merged**